lib/core/base/cache.js

```diff
@@ -13,9 +13,24 @@ const cache = {
   /**
    * Retrieve an item from the cache.
    * @param {String} key - Name of the key the value was stored as.
+   * @param {*} defaultValue - Default value to set if there is a cache miss. Functions are evaluated before caching. To override a value already saved, use `set()`.
    * @returns {*} The item stored
    */
-  get(key) {
+  get(key, ...args) {
+    const defaultValue = args[0];
+    const hasDefaultValue = args.length === 1;
```

Comment:
Why do we need to rely on arity here? This smells like magic.

Reply:
Also, what if someone accidentally passes 2 parameters to this function?

Reply:
Since `undefined` is treated as a value that could be cached, we can't just check for `undefined` of the second parameter. 

The best way I can think of doing it is to look at the length of arguments passed by the caller by spread or `arguments.length`.

The params have had a couple different ways of getting defined suggested. Personally, I'm not a huge fan of using the spread of args as the second parameter since only the first value would be used. I'll toss another one and see if that floats higher in the water for us:

```suggestion
    get(key, defaultValue) {
        const hasDefaultValue = arguments.length === 2;
```
IMO, a named parameter instead of `...args` addresses most of the magic smell. The other thing that seems more obvious to me is to make the second parameter an options object. `options = { default: null }`

If someone sends a second argument in accidentally, they would be setting the cache for that key if the key was not previously defined. It does not override an existing cache value though.

We could also leave `cache.get()` untouched and add Wilco's requested logic to a new method, something like `cache.getWithDefaultValue()`.

```suggestion
    getWithDefaultValue(key, defaultValue) {
        if( !get(key) ){
            set(key, defaultValue); 
            //plus function detection with .call()
        };
        return get(key);
```

Reply:
Why would we need to cache `undefined`?

Reply:
My understanding is that Axe throws around `undefined` as a meaningful value by convention, and we may want to be caching that return. `cache.set()` will let you do so currently.

Reply:
axe-core uses `undefined` to denote an incomplete check. I can't think of any reason we'd cache that though.

Unless we have an actual use case for doing `cache.set('foo', undefined)`, I do not think we should support it.

Reply:
Tagging @WilcoFiers and @straker

Reply:
@stephenmathieson There are a few functions in axe-core where `undefined` is a valid cache value, such as [`is-modal-open](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/is-modal-open.js#L100-L101). The point of the cache is to denote an operation (potentially expensive) that we do not want to do again. What the operation returns is not important.

Reply:
What is the difference between not setting anything in the cache and setting `undefined`? I don't understand why we'd ever need this.

Reply:
Setting anything in the cache means we've performed an operation and do not wish to do it again. If it helps, think of it as memoizing a function, but instead of caching the value for a set of parameters, we can cache the value for the entire run. It doesn't matter if the memoized function returned `undefined` (or anything really), the function will not be run again for the same set of parameters.

Reply:
Perhaps I don't understand the code, but it looks like the function will run again if `undefined` is cached based on these lines: https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/is-modal-open.js#L37-L39

Reply:
Thanks for pointing that out, looks like a bug in the `is-modal-open` function.

Reply:
I still don't believe there's a reason to cache `undefined`.